### DISCUSSION TITLE
Can I run a JavaScript script and a Bash script in a composite GitHub Action?

### DISCUSSION THREAD

**Discussion by PaulRBerg** [community/community#26248] [Actions]
<p>I’m looking to create my own <a href="https://docs.github.com/en/actions/creating-actions/about-custom-actions" rel="noopener nofollow ugc">custom action</a>, and the task that I’m working on would be easiest to implement by doing some work in JavaScript and some other work in a Bash script.</p>
<p>Can I do this with a <a href="https://docs.github.com/en/actions/creating-actions/creating-a-composite-action" rel="noopener nofollow ugc">composite action</a>, without creating another action for the JavaScript script?</p>

**Comment by jsoref** [comment, marked as answer]
<p>Sure.</p>
<p>You could use:</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/actions/github-script">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/actions/github-script" target="_blank" rel="noopener">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/2119212/181096432-2256cc6a-7807-4a3c-8e70-5f82430e5ef4.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/3/1/3104b527aed24da278ab8b480d9b7c6f069a6e9d_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/3/1/3104b527aed24da278ab8b480d9b7c6f069a6e9d_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/3/1/3104b527aed24da278ab8b480d9b7c6f069a6e9d.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/3/1/3104b527aed24da278ab8b480d9b7c6f069a6e9d_2_10x10.png"></div>

<h3><a href="https://github.com/actions/github-script" target="_blank" rel="noopener">GitHub - actions/github-script: Write workflows scripting the GitHub API in...</a></h3>

  <p>Write workflows scripting the GitHub API in JavaScript - GitHub - actions/github-script: Write workflows scripting the GitHub API in JavaScript</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Or you could just use <code>node</code>…</p>

**Reply by PaulRBerg** [reply]
@jsoref, I have just spent a few hours trying to integrate this action in my set-up, but I just couldn't make it work. I have added this YAML code to my composite action:

```yaml
uses: "actions/github-script@v6"
with:
  script: |
    const script = require("./dist/index.js")
    script({core})
```

But I am getting this error when I run it:

```text
Error: Cannot find module '/home/runner/work/v2-core/v2-core/dist/index.js'
Require stack:
- /home/runner/work/_actions/actions/github-script/v6/dist/index.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.apply (/home/runner/work/_actions/actions/github-script/v6/dist/index.js:13400:27)
    at eval (eval at callAsyncFunction (/home/runner/work/_actions/actions/github-script/v6/dist/index.js:13356:16), <anonymous>:3:16)
    at callAsyncFunction (/home/runner/work/_actions/actions/github-script/v6/dist/index.js:13357:12)
    at main (/home/runner/work/_actions/actions/github-script/v6/dist/index.js:13452:26)
    at Module.858 (/home/runner/work/_actions/actions/github-script/v6/dist/index.js:13429:1)
    at __webpack_require__ (/home/runner/work/_actions/actions/github-script/v6/dist/index.js:[24](https://github.com/sablierhq/v2-core/actions/runs/3861794209/jobs/6583130559#step:3:26):[31](https://github.com/sablierhq/v2-core/actions/runs/3861794209/jobs/6583130559#step:3:33)) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/runner/work/_actions/actions/github-script/v6/dist/index.js'
  ]
}
```

Do you have any idea what am I doing wrong?

**Reply by jsoref** [reply]
https://cs.github.com/?scopeName=All+repos&scope=&q=%22actions%2Fgithub-script%40v6%22+require suggests:
 https://github.com/mlflow/mlflow/actions/runs/3861219273/jobs/6582062172
https://github.com/mlflow/mlflow/blob/92aa9fac426fde9cde0fbaaa4a0ce71468da7cbd/.github/workflows/advice.yml#L15-L19
```js
      - uses: actions/github-script@v6
        with:
          script: |
            const script = require(
              `${process.env.GITHUB_WORKSPACE}/.github/workflows/advice.js`
```

**Reply by PaulRBerg** [reply]
Thanks @jsoref, though I don't think this addresses my use case. I am interested in building a composite action that is then run by another action.

Looks like it is not possible to bundle up JavaScript code unless the action is JavaScript only.

**Reply by jsoref** [reply]
Shrug. You can definitely have a JavaScript action run a shell. – That's how check-spelling worked before they introduced composite actions.

A composite action could certainly run your composite action. And I'm pretty sure that your action should be able to include a uses step like the one above.

**Reply by PaulRBerg** [reply]
> You can definitely have a JavaScript action run a shell

It's the other way around - I would like a composite action to run a JavaScript action - or, more simply put, run a JavaScript file.

**Reply by jsoref** [reply]
This really should do what you need: https://github.com/orgs/community/discussions/26248#discussioncomment-4622511

**Reply by Jimmy-TheOz** [reply]
> @jsoref, I have just spent a few hours trying to integrate this action in my set-up, but I just couldn't make it work. I have added this YAML code to my composite action:
> 
> ```yaml
> uses: "actions/github-script@v6"
> with:
>   script: |
>     const script = require("./dist/index.js")
>     script({core})
> ```
> 
> But I am getting this error when I run it:
> 
> ```
> Error: Cannot find module '/home/runner/work/v2-core/v2-core/dist/index.js'
> Require stack:
> - /home/runner/work/_actions/actions/github-script/v6/dist/index.js
>     at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
>     at Function.Module._load (node:internal/modules/cjs/loader:778:27)
>     at Module.require (node:internal/modules/cjs/loader:1005:19)
>     at require (node:internal/modules/cjs/helpers:102:18)
>     at Object.apply (/home/runner/work/_actions/actions/github-script/v6/dist/index.js:13400:27)
>     at eval (eval at callAsyncFunction (/home/runner/work/_actions/actions/github-script/v6/dist/index.js:13356:16), <anonymous>:3:16)
>     at callAsyncFunction (/home/runner/work/_actions/actions/github-script/v6/dist/index.js:13357:12)
>     at main (/home/runner/work/_actions/actions/github-script/v6/dist/index.js:13452:26)
>     at Module.858 (/home/runner/work/_actions/actions/github-script/v6/dist/index.js:13429:1)
>     at __webpack_require__ (/home/runner/work/_actions/actions/github-script/v6/dist/index.js:[24](https://github.com/sablierhq/v2-core/actions/runs/3861794209/jobs/6583130559#step:3:26):[31](https://github.com/sablierhq/v2-core/actions/runs/3861794209/jobs/6583130559#step:3:33)) {
>   code: 'MODULE_NOT_FOUND',
>   requireStack: [
>     '/home/runner/work/_actions/actions/github-script/v6/dist/index.js'
>   ]
> }
> ```
> 
> Do you have any idea what am I doing wrong?

@PaulRBerg did you figured it out how to do this ? 
I'm in the same situation, i tried node ${{ github.action_path }}/dist/index.js but doesn't work neither

**Reply by PaulRBerg** [reply]
@Jimmy-TheOz nope, though I did get a little bit of extra advice here:

https://github.com/actions/github-script/issues/322

**Reply by jsoref** [reply]
Does the file exist?

Before that step, can you run a `find /home/runner/work/v2-core/v2-core/`?

**Reply by PaulRBerg** [reply]
@jsoref thank you for your help but I think we're talking about different scenarios. The file exists in the third-party GitHub Action - though that action is invoked by the `v2-core` repo. I might be misunderstanding something here though - it might be that you cannot call arbitrary JavaScript code from a third-party action unless that action packages the code as a JavaScript action.

**Reply by jsoref** [reply]
@PaulRBerg ??

@Jimmy-TheOz wrote:

```yaml
uses: "actions/github-script@v6"
with:
  script: |
    const script = require("./dist/index.js")
    script({core})
```

With this error:
```
Error: Cannot find module '/home/runner/work/v2-core/v2-core/dist/index.js'
```

The action itself (`actions/github-script@v6`) lives in:
```
    '/home/runner/work/_actions/actions/github-script/v6/dist/index.js'
```

**Reply by PaulRBerg** [reply]
Oh, sorry @jsoref, I thought you were talking to me. Sorry for the spam.

**Reply by Jimmy-TheOz** [reply]
@PaulRBerg FYI, I ended up by doing this : 
```
    - name: "Run script"
      run: node ${{ github.action_path }}/dist/index.js
      shell: bash
```

this important part is the ${{ github.action_path }}, hope this will help you ;)

**Reply by intc-hharshtk** [reply]
Is there any documentation somewhere for the same??

**Comment by PaulRBerg** [comment]
<p>Oh, <code>github-script</code> is exactly what I was looking for. Many thanks!</p>
